compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java

```diff
@@ -484,37 +481,134 @@ else if (atSet(Precedence.POSTFIX.getOperations())) {
+
+            return true;
         }
         else if (at(LPAR)) {
             parseValueArgumentList();
             parseCallWithClosure();
+
+            return true;
         }
-        else if (at(LT)) {
-            PsiBuilder.Marker typeArgumentList = mark();
-            if (myKotlinParsing.tryParseTypeArgumentList(TYPE_ARGUMENT_LIST_STOPPERS)) {
-                typeArgumentList.done(TYPE_ARGUMENT_LIST);
-                if (!myBuilder.newlineBeforeCurrentToken() && at(LPAR)) parseValueArgumentList();
-                parseCallWithClosure();
-            }
-            else {
+        else if (parseCallWithClosure()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean parseCallSuffixTypeArgumentList() {
+        PsiBuilder.Marker typeArgumentList = mark();
+
+        KotlinParsing.TypeArgumentListKind kind = myKotlinParsing.tryParseTypeArgumentList(/* strict */false);
+
+        // If the parsed section did not fulfill the criteria of a type argument list or
+        // if there's no indication of an intended call suffix, drop the type argument list.
+        if (kind == KotlinParsing.TypeArgumentListKind.NONE || !isAtTypeArgumentListIndicator()) {
+            typeArgumentList.rollbackTo();
+
+            return false;
+        }
+
+        // If the parsed section could have been a faulty type argument type list, check
+        // whether it could have also been parsed as the right-hand side of a comparison
+        // expression correctly. If so, we have never had a faulty type argument list in
+        // the first place, and therefor it must be dropped.
+        if (kind == KotlinParsing.TypeArgumentListKind.FAULTY_TYPE_ARGUMENT_LIST) {
+            // For better error highlighting we only try to produce a faulty type argument list, if
+            // there are strong indicators. Otherwise, we would lean too much towards call suffixes
+            // in cases like when typing "x < 0 && 1 > ", where the last element hasn't been typed
+            // yet.
+            if (!isAtStrongTypeArgumentListIndicator()) {
                 typeArgumentList.rollbackTo();
+
                 return false;
             }
+
+            typeArgumentList.rollbackTo();
+
+            if (isAtConditionalExpression()) {
+                return false;
+            }
+
+            typeArgumentList = mark();
+
+            myKotlinParsing.tryParseTypeArgumentList(/* strict */true);
         }
-        else {
-            return false;
-        }
+
+        // At this point we're sure that it's either a correct type argument list followed
+        // by some indication of a call suffix or that it's faulty type argument list that
+        // is definitely not a correct right-hand side of an expression and is followed by
+        // a strong indication of a call suffix.
+
+        typeArgumentList.done(TYPE_ARGUMENT_LIST);
 
         return true;
     }
 
+    private boolean isAtTypeArgumentListIndicator() {
```

Comment:
I would rename it to something like `isAtTypeArgumentListSuccessorIndicator` because otherwise one might think from its name that it's about the beginning of argument list.

The same may apply to other similar named-functions below and to the token set fields, too.